src/main/kotlin/org/rust/lang/core/types/RustTypeParameterType.kt

```diff
@@ -16,6 +16,8 @@ class RustTypeParameterType(val parameter: RustTypeParamElement) : RustTypeBase(
     override fun getNonStaticMethodsIn(project: Project): Sequence<RustFnElement> =
         getTraitsImplementedIn(project).flatMap { it.traitMethodMemberList.asSequence() }
 
+    override fun substitute(map: Map<RustTypeParameterType, RustType>): RustType = map[this] ?: this
+
```

Comment:
We should not map over `RustTypeParameterType`, b/c we're misusing the abstraction -- we're mapping over the _name_ (and implicit position) of the parameter, not against its type.

Let's replaced `RustTypeParameterType` here with abstract `RustTypeParameterTag` of which every `RustTypeParameter` would've got a _unique_ one.